### DISCUSSION TITLE
📔【操作系统】陷阱、中断、异常、信号

### DISCUSSION THREAD

**Discussion by imageslr** [imageslr/blog#86] [Announcements]
https://imageslr.github.io/2020/07/09/trap-interrupt-exception.html 

 前言：异常控制流 每个进程对应的程序文件由一条一条的指令组成。进程在执行的时候，会将程序文件加载到进程的内存空间中，这些指令在内存空间中是相邻的。进程会通过调整程序计数器 PC 的值，一条一条地执行指令。我们将进程执行的指令序列叫做处理器的「控制流」。

**Comment by nengqi** [comment]
陷阱也被称为软中断，好像有点问题
系统调用应该就是软中断
而陷阱trap应该是exception的一个子类

**Comment by imageslr** [comment]
@nengqi 你的意思是，陷阱 trap 属于一个异常（不符合预期），而不是一个软中断（符合预期）？

可以举几个 exception 中属于 trap 的例子吗？

**Comment by nengqi** [comment]
https://stackoverflow.com/questions/56877451/difference-between-trap-and-exceptions
感觉是概念问题


齐能
***@***.***




------------------ 原始邮件 ------------------
From: Images。 ***@***.***&gt;
发送时间: 10/26/2021, 18:39
To: imageslr/blog ***@***.***&gt;
抄送: Qi Neng ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Reply:[imageslr/blog] ��【操作系统】陷阱、中断、异常、信号 (#44)





 
@nengqi 你的意思是，陷阱 trap 属于一个异常（不符合预期），而不是一个软中断（符合预期）？
 
可以举几个 exception 中属于 trap 的例子吗？

**Comment by imageslr** [comment]
> @nengqi
> https://stackoverflow.com/questions/56877451/difference-between-trap-and-exceptions
> 感觉是概念问题

赞。感觉可以这样理解：如果是符合预期的“好”的异常 exception，就是陷阱 trap；如果是不符合预期的“坏”的异常，就是故障 abort。这个概念在《CSAPP》上有描述，我之后再看一下。

**Comment by Vector6662** [comment]
我觉得是翻译的问题，文中所谓的“坏的异常”应该指的是为error，而“异常控制流”的异常exception，我觉得翻译为例外更好
大佬的这篇一章下的[C 访问空指针会不会陷入内核态](https://imageslr.com/2020/07/07/user-mode-kernel-mode.html#c-%E8%AE%BF%E9%97%AE%E7%A9%BA%E6%8C%87%E9%92%88%E4%BC%9A%E4%B8%8D%E4%BC%9A%E9%99%B7%E5%85%A5%E5%86%85%E6%A0%B8%E6%80%81) 引用的StackOverflow下的一句话：
> the CPU raises a page fault **error** which traps into a predefined point at the kernel, the kernel examines what happened, and reacts accordingly

用的就是page fault **error**

**Comment by imageslr** [comment]
@Vector6662 有道理。这里的名词 interrupt、trap、abort、error、exception 当时也让我纠结了很久。

在第四小节中，我其实是没有明确区分”异常 exception”、“错误 error” 和 “故障 abort”。当时这么写的：
> 注意这里的“异常”和上文一开始说的“异常”的区别。上文的“异常”是一个通用的术语，表示因为某些事件/操作而引起的控制流的改变，包括陷阱、中断和异常。这里的“异常”特指因为执行当前指令而产生的错误情况，比如除法异常、缺页异常等。有些书上为了区分，也将这类“异常”称为“故障”。

所以全文都是在说“异常控制流”，这里的语义和你表达的“例外”是一样的。

然后如果仅专注于第四小节的话，我把“异常”“错误情况”“故障”都统称为异常了。严谨点来说，那里用“错误 error”或者“故障 abort”更合适一些。

**Comment by Vector6662** [comment]
参考资料里边的《Linux系统调用指南》链接挂了，请问能给个新的链接吗？

**Comment by imageslr** [comment]
> 参考资料里边的《Linux系统调用指南》链接挂了，请问能给个新的链接吗？

@Vector6662  已更新
